MapboxCoreNavigation/RouteStep.swift

```diff
@@ -54,23 +56,16 @@ extension RouteStep {
     }
     
     /**
-     Returns a tunnel slice for the current route step coordinates
-     */
-    var tunnelSlice: Polyline? {
-        guard let coordinates = coordinates, let intersectionBounds = tunnelIntersectionBounds else { return nil }
-        return Polyline(coordinates).sliced(from: intersectionBounds.entry.location, to: intersectionBounds.exit.location)
-    }
-    
-    /**
-     Returns the entry and exit intersection bounds of the tunnel on the current route step
+     Returns an array of the entry and exit intersection bounds of the tunnel on the current route step
      */
-    var tunnelIntersectionBounds: (entry: Intersection, exit: Intersection)? {
-        guard let intersections = intersections, containsTunnel else { return nil }
-        for i in 0..<(intersections.count - 1) where intersections.count > 1 {
-            if intersections[i].outletRoadClasses == .tunnel {
-                return (entry: intersections[i], exit: intersections[i+1])
+    var tunnelIntersectionsBounds: [IntersectionBounds]? {
+        guard let intersections = intersections, intersections.count > 1, containsTunnel else { return nil }
+        var intersectionBounds = [IntersectionBounds]()
+        for i in 0..<intersections.count {
+            if let outletRoadClasses = intersections[i].outletRoadClasses, outletRoadClasses.contains(.tunnel) && i < intersections.count - 1 {
+                intersectionBounds.append((entry: intersections[i], exit: intersections[i+1]))
```

Comment:
If we do keep this method around, I think it should look like either of the suggestions in https://github.com/mapbox/mapbox-navigation-ios/pull/1127#discussion_r169501520, because manual index management has been a cause for difficult-to-debug out-of-bounds errors in the past in this library.

Reply:
@1ec5 I wanted to go with code readability and simplicity. Where do you sense any difficult-to-debug out-of-bounds errors from this condition?  

Follow up: Perhaps, if you were to compose a self documented functional programming, which can be easily maintained (modified)? Also, wouldn't it be less optimal given we need to do series of operations -- such as `dropLast()`, `contains`, `index(where:)`, but just to mention a few.

Reply:
> I wanted to go with code readability and simplicity. Where do you sense any difficult-to-debug out-of-bounds errors from this condition?

In this case, I’m not concerned that the code would have a difficult-to-debug bug in it, but that it’s easier for an off-by-one error to creep in when reviewing. In general, it’s a best practice in Swift to avoid manual index management, just as it’s best to avoid manual memory access. The functional approach also happens to be more of a literate programming style in this case.

> wouldn't the proposed way be less optimal given we have to do series of chained operations -- such as `dropLast()`, `contains`, `index(where:)`?

The second approach in https://github.com/mapbox/mapbox-navigation-ios/pull/1127#discussion_r169501520 is just as efficient as this code: `dropLast()` returns a lazy array, so there’s no performance penalty, and `index(where:)` and `index(after:)` are both _O_(1) operations. `contains(_:)` is necessary because RoadClass is a bitmask. (The current code uses it too.)

Reply:
Since we now need to iterate over all the intersections, even after finding a tunnel, the `zip` approach looks more desirable:

```swift
// Pair each intersection with its successor.
let intersectionPairs = zip(intersections, intersections.prefix(from: 1))
// Return the first pair of intersections that’s in a tunnel.
return intersectionPairs.first { $0.0.outletRoadClasses?.contains(.tunnel) ?? false }.map { IntersectionBounds($0.0, $0.1, coordinates) }
```